src/System.Device.Gpio/System/Device/Gpio/Drivers/Libgpiod/LibGpiodDriverFactory.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Device.Gpio.System.Device.Gpio.Drivers.Libgpiod.V1;
+using System.Device.Gpio.System.Device.Gpio.Drivers.Libgpiod.V2;
+using System.IO;
+using System.Linq;
+
+namespace System.Device.Gpio.System.Device.Gpio.Drivers.Libgpiod;
+
+internal static class LibGpiodDriverFactory
+{
+    private const int _defaultDriverVersion = 1;
+
+    private const string _librarySearchPattern = "libgpiod.so*";
+    private const string _v0 = "libgpiod.so.0";
+    private const string _v1_0 = "libgpiod.so.1";
+    private const string _v1_1 = "libgpiod.so.2";
+    private const string _v2 = "libgpiod.so.3";
+
+    private static readonly string[] _librarySearchPaths = { "/lib", "/usr/lib", "/usr/local/lib" };
+
+    private static readonly object _lockObject = new();
+    private static int? _driverVersionToLoad;
+
+    /// <summary>
+    /// Tries with best effort to find installed libgpiod libraries and loads the latest version.
+    /// <see href="https://learn.microsoft.com/en-us/dotnet/standard/native-interop/native-library-loading#custom-import-resolver"/>
+    /// </summary>
+    /// <remarks> On NET Core 3.1 or NET5+ SetDllImportResolver would be a better solution</remarks>
+    public static GpioDriver Create(int chipNumber)
+    {
+        lock (_lockObject)
```

Comment:
This is a static constructor on a static class, I don't think that we need to introduce locks go guarantee for this to only run once, unless I'm missing something.

Reply:
duh! clearly missing something haha, this is not a constructor but instead a Create method. It does make it feel expensive though to have to use a lock everytime someone calls Create. Can we consider having a static constructor that essentially finds out which version to use and cache t hat, and then all subsequent calls to Create can avoid locking?

Reply:
Yeah that makes sense. Let's see on other comments on how we will move on with this factory